frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts

```diff
@@ -251,5 +284,249 @@ void main() {
+uniform float viewMode;
+uniform float alpha;
+uniform bool renderBucketIndices;
+uniform vec3 bboxMin;
+uniform vec3 bboxMax;
+uniform vec3 globalPosition;
+uniform vec3 activeSegmentPosition;
+uniform float zoomValue;
+uniform bool useBilinearFiltering;
+uniform vec3 globalMousePosition;
+uniform bool isMouseInCanvas;
+uniform float brushSizeInPixel;
+uniform float planeID;
+uniform vec3 addressSpaceDimensions;
+uniform vec4 hoveredSegmentIdLow;
+uniform vec4 hoveredSegmentIdHigh;
+
+const float bucketWidth = <%= bucketWidth %>;
+const float bucketSize = <%= bucketSize %>;
+const float l_texture_width = <%= l_texture_width %>;
+
+
+// For some reason, taking the dataset scale from the uniform results in imprecise
+// rendering of the brush circle (and issues in the arbitrary modes). That's why it
+// is directly inserted into the source via templating.
+const vec3 datasetScale = <%= formatVector3AsVec3(datasetScale) %>;
+
+const vec4 fallbackGray = vec4(0.5, 0.5, 0.5, 1.0);
+
+${compileShader(
+  inverse,
+  div,
+  isNan,
+  isFlightMode,
+  transDim,
+  getAbsoluteCoords,
+  getWorldCoordUVW,
+  isOutsideOfBoundingBox,
+  getMaybeFilteredColorOrFallback,
+  hasSegmentation ? getSegmentationId : null,
+  getResolution,
+)}
+
+
+float PLANE_WIDTH = ${formatNumberAsGLSLFloat(Constants.VIEWPORT_WIDTH)};
+float PLANE_SUBDIVISION = ${formatNumberAsGLSLFloat(PLANE_SUBDIVISION)};
+
+void main() {
+  vUv = uv;
+  modelCoord = vec4(position, 1.0);
+  savedModelMatrix = modelMatrix;
+  mat4 modelInv = inverseMatrix(modelMatrix);
+  worldCoord = modelMatrix * vec4(position, 1.0);
+
+  gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
+
+
+  // Remember, the top of the viewport has Y=1 whereas the left has X=-1.
+  vec3 worldCoordTopLeft     = transDim((modelMatrix * vec4(vec3(-PLANE_WIDTH/2.,  PLANE_WIDTH/2., 0.), 1.)).xyz);
+  vec3 worldCoordBottomRight = transDim((modelMatrix * vec4(vec3( PLANE_WIDTH/2., -PLANE_WIDTH/2., 0.), 1.)).xyz);
+
+  // The following code ensures that the vertices are aligned with the bucket borders
+  // of the currently rendered magnification.
+  // In general, each vertex is floored to the next lower multiple of a bucket's extent.
+  // Exceptions are the following:
+  // - the first and the last vertex isn't moved so that the plane keeps its original extent
+  // - the second last vertex is moved so that it's positioned at the beginning of the last
+  //   rendered bucket. If this isn't handled additionally, the second last vertex might be
+  //   still correct in most cases, but at certain positions, the second last vertex is
+  //   already in the second last bucket and would be moved to far to the left.
+  //   Example:
+  //   *------|------|------|----a-|---*
+  //   ^ first vertex
+  //          ^ second vertex
+  //                   ...             ^ last vertex
+  //                             ^ if a is the second last vertex, it needs to be moved to the right
```

Comment:
Doesn't the same issue exist for the third last vertex then, sometimes. That it maybe would need to be moved to the right, as otherwise the second last bucket would not be covered? (and so on...)